core/modules/user/user.module

```diff
@@ -879,49 +878,39 @@ function user_account_form_validate($form, &$form_state) {
-    if ($name == '' && $hide_username_on_registration && empty($account->uid)) {
-      $username_generation_method = config_get('system.core', 'username_generation_method');
-      switch ($username_generation_method) {
-        case 0:
-          $name = $mail;
-          break;
-        case 1:
-          $name = user_generate_username($mail);
-          break;
-        case 2:
-          $name = user_generate_username();
-          break;
-      }
-      $form_state['values']['name'] = $name;
-    }
-    if ($error = user_validate_name($name)) {
-      form_set_error('name', $error);
-    }
-    // Cast the user ID as an integer. It might have been set to NULL, which
-    // could lead to unexpected results.
-    else {
-      $name_taken = (bool) db_select('users')
-        ->fields('users', array('uid'))
-        ->condition('uid', (int) $account->uid, '<>')
-        ->condition('name', db_like($name), 'LIKE')
-        ->range(0, 1)
-        ->execute()
-        ->fetchField();
-
-      if ($name_taken) {
-        form_set_error('name', t('The name %name is already taken.', array('%name' => $name)));
+    // Admins have the option to generate the username.
+    if (!config_get('system.core', 'generate_usernames')){
+      if ($error = user_validate_name($name)) {
+        form_set_error('name', $error);
       }
-      // Check whether the user name provided is an email address. If so, make
-      // sure it matches the mail value.
-      if (config('system.core')->get('user_email_match') && (valid_email_address($name))) {
-        if ($name !== $mail) {
-          form_set_error('name', t('An email address was provided as a username, but does not match the account email address.'));
+      // Cast the user ID as an integer. It might have been set to NULL, which
+      // could lead to unexpected results.
+      else {
+        $name_taken = (bool) db_select('users')
+          ->fields('users', array('uid'))
+          ->condition('uid', (int) $account->uid, '<>')
+          ->condition('name', db_like($name), 'LIKE')
+          ->range(0, 1)
+          ->execute()
+          ->fetchField();
+
+        if ($name_taken) {
+          form_set_error('name', t('The name %name is already taken.', array('%name' => $name)));
+        }
+        // Check whether the user name provided is an email address. If so, make
+        // sure it matches the mail value.
+        if (config('system.core')->get('user_email_match') && (valid_email_address($name))) {
+          if ($name !== $mail) {
+            form_set_error('name', t('An email address was provided as a username, but does not match the account email address.'));
+          }
         }
       }
     }
+    else {
+      // Here we can use user_password() to generate a safe random string.
+      // This username will be updated t use the UID in hook_user_insert().
+      $form_state['values']['name'] = 'account-' . user_password();
```

Comment:
if this is temporary, use a more unguessable string than 'account-'. its not unlikely that a real user would choose that as the prefix of his username. 'tmpacctstrng'?